### ISSUE TITLE
Edit / add synonyms to model

### ISSUE THREAD

**Issue by troyel** [otykier/TabularEditor#257]
Any possibility of adding support to edit "Synonyms" for PBI models? I have tried adding synonyms and see no change in the Tabular Editor. Would be fantastic to do this from the editor as well.

**Comment by otykier** [comment]
I’d like to support this, but last time I checked, Power BI was using a weird XML-based format to store the linguistic metadata embedded within the Model metadata. The AMO libs used by Tabular Editor to manipulate the TOM does not support this format. It does expose a set of properties and classes related to linguistic metadata, but it does not seem to work with Power BI files. Also, adding linguistic metadata to SSAS Tabular models is not supported by Power BI (as Q&A is not supported with Live connections), so it’s quite a mess at the moment, and I’m not really sure what to make of it.

**Comment by marcosqlbi** [comment]
I'd suggest to wait - long term, they should implement it in JSON or at least make it supported. Doing something now means trying to reverse engineer their format / unless it is documented somewhere...

**Comment by troyel** [comment]
I'm not sure if it is out of preview yet but Q&A is supported for SSAS live connection (https://docs.microsoft.com/en-us/power-bi/service-q-and-a-direct-query)

**Comment by otykier** [comment]
Thanks - I was not aware that Q&A is now supported for live connections. I will investigate whether the [LinguisticMetadata property](https://docs.microsoft.com/en-us/dotnet/api/microsoft.analysisservices.tabular.linguisticmetadata?view=analysisservices-dotnet) of the TOM actually works with Power BI Q&A, because in that case, it most definitely makes sense to let you edit these in Tabular Editor.

**Comment by Yasston** [comment]
I'd also be interested in this feature. We develop tabular models deployed on Azure AS. Q&A with Live connection does work pretty well in PBI, we are only missing the possibility of defining synonyms :) Even if not intensively used in production, Q&A is quite good for a nice "wow effect" with some of our management, and it does look promising.
Available for testing if needed. 
Many thanks !

**Comment by benito2313** [comment]
Hi,

We are also interested in this feature.
Is there any change in this?

Kind regards,

Benito2313

**Comment by otykier** [comment]
Unfortunately, I haven't been able to find any documentation on how to use the LinguisticMetadata property. I could certainly expose the property within Tabular Editor, but you'd still need to figure out how the JSON or XML content should be changed in order to add synonyms, etc. Today, you can access the LinguisticMetadata property through scripting only:

```csharp
// WARNING: When we're accessing the underlying TOM object using .GetMetadataObject(), we're circumventing
// Tabular Editor's undo/redo framework. Any changes made to the tomCulture object in this script, can not
// be undone!
var tomCulture = Model.Cultures["da-DK"].GetMetadataObject() as TOM.Culture;
tomCulture.LinguisticMetadata.Content.Output();
```

![image](https://user-images.githubusercontent.com/8976200/100984724-5e8c3b80-354b-11eb-88a3-74ef5a790245.png)


Would it suffice if Tabular Editor exposed the LinguisticMetadata properties for a specific culture in the Property Grid? This would allow you to edit the XML or JSON content as a string. I'd love to provide a better experience, but this is not possible until there's full documentation (or even better, an API!) for the XML/JSON formats!

**Comment by otykier** [comment]
[Tabular Editor 2.13.2](https://github.com/otykier/TabularEditor/releases/latest) now exposes the linguistic metadata content in the property grid. You can modify the value in the grid, or you can use scripting to set the `Content` property to any XML or Json value:

![image](https://user-images.githubusercontent.com/8976200/101163230-a7b7ba80-3633-11eb-8284-a17da3d6d1e3.png)

**Comment by benito2313** [comment]
Hi,

Thnx for adding this feature.

Is there anything that can be done by adding a field called synonyms ? just like with Annotations, that we can add multiple synonyms.
![image](https://user-images.githubusercontent.com/3872421/112119267-81effa80-8bbd-11eb-8c60-17077b59e950.png)

And also could u add it to the translations tab:
![image](https://user-images.githubusercontent.com/3872421/112119569-caa7b380-8bbd-11eb-8c95-acdbbbcfc0bf.png)
Because you would also like to translate this.

There is some information about it, but i dont know if its enough
https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-ssas-t/dd77faf6-2bef-457e-b7a0-3add19aaab9e
https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-ssas-t/f8924a45-70da-496a-947a-84b8d5beaae6

I also have an output from a pbix file with the formating:
![image](https://user-images.githubusercontent.com/3872421/112122676-d183f580-8bc0-11eb-8f99-0edffd2b5bd6.png)
I know this is attached to an translation, but you would like to store the content in a main place just like you do with the description and name and so on.

And when i use metadata translator i would like to also retrieve those synonyms to also translate them.

Kind regards,

Ben

**Comment by otykier** [comment]
[Version 2.15.2](https://github.com/otykier/TabularEditor/releases/latest) now attempts to parse the culture `Content` property, in order to show a comma-separated list of synonyms defined for each table, column, measure, hierarchy and level.

This works as of the March 2021 release of Desktop, but there are no guarantees that this will continue working. Specifically, since the Linguistic Schema Definition Language (which is what Microsoft call the JSON data in the `Content` property) is not really documented, Microsoft could basically change the structure in a later version of Desktop, potentially breaking Tabular Editor's support for adding/editing synonyms. You have been warned! :-)

**Comment by benito2313** [comment]
Hi, 
Thnx for the quick update :)
But where can i add a synonym in the model in tabular editor? I dont want to do this in powerbi desktop, this because we work with AS.

Kind regards,
Ben

**Comment by otykier** [comment]
@benito2313 I'm not sure if synonyms defined in an SSAS/Azure AS model are supported by Power BI at all. Do you have any links that indicates that this should be the case?

 Tabular Editor will let you add/edit synonyms on objects once a JSON-based LSDL schema has been specified on the `Content` property on your culture. For a Power BI Desktop model, this schema would be specified by Power BI, with Tabular Editor only adding/removing synonym terms (but there's a lot of other stuff going on in the LSDL, which Tabular Editor doesn't touch).
 
 So in your case, you can try to add the following empty LSDL schema as the `Content` of your culture (I guess Q&A only works with en-US at the moment):
 
 ```json
 {"Version":"1.2.0","Language":"en-US","DynamicImprovement":"None","Entities":{}}
 ```
 
![image](https://user-images.githubusercontent.com/8976200/112494319-38e9a300-8d83-11eb-8c06-4128c068b49d.png)

 Once you add this string, Tabular Editor will display a "Synonyms" property on all objects that can have synonyms assigned:
 
![image](https://user-images.githubusercontent.com/8976200/112494475-5dde1600-8d83-11eb-8a8b-fd6961ab4732.png)

Modifying the "Synonyms" value will basically just update the LSDL schema to include the synonyms:

![image](https://user-images.githubusercontent.com/8976200/112494726-9382ff00-8d83-11eb-957b-70209ebeeb21.png)

But again - there are no guarantees that this actually works, since I couldn't find anywhere that says synonyms are supported for live connections against SSAS/Azure AS. But feel free to try it and kindly let us know the result. Thanks!

**Comment by benito2313** [comment]
This look great i'm going to check it out if it works in pbi.

In the mean while i have this link https://powerbi.microsoft.com/en-us/blog/editing-q-a-linguistic-schemas/
Also at the end of the blog there is a download link to the schema.

![image](https://user-images.githubusercontent.com/3872421/112542303-b1b52300-8db4-11eb-8176-6bfd442f86cb.png)

Please let me know if it was helpful, or that i need to search further for you.

Kind regards,

Ben

**Comment by benito2313** [comment]
Hi Daniel,

I was wondering if the schema is updated for synonyms in tabular editor?
As @Aaron-Meyers suggested in https://github.com/microsoft/Analysis-Services/issues/100 :
The TOM LinguisticMetadata schema can be found at aka.ms/lsdlschema or json.schemastore.org/lsdlschema

I saw there is a new version (3.3.) and I was wondering if it could be updated in TE?
https://json.schemastore.org/lsdlschema-3.3.json

I have also found more documentation from Microsoft that was added/modified on 30-07-2021.
https://docs.microsoft.com/en-us/power-bi/natural-language/q-and-a-tooling-advanced#prerequisites 
It’s on line “Download sample .yaml and .pbix files” , at this part of the post you can download a zip with three files
- Linguistic Schema Spec.docx
- - in this file the schema is fully explained!
- Sample PBIX
- Sample yaml 
The direct link to the zip is https://go.microsoft.com/fwlink/?linkid=871858 .

So to summarize we have the schemas and the explanation of the schema.
Is this something that you can use to add more functionality to TE?

Kind regards,

Ben

**Comment by otykier** [comment]
Hi @benito2313 

There will not be better linguistic schema support in Tabular Editor, as long as Microsoft does not provide an API for programatically accessing the schema. Implementing my own API based on the linguistic schema specs would be a tremendous effort, and it would have to me maintained every time there is an update of the schema (which happens often!).

This is in contrast to the TOM itself, where [Microsoft provided an excellent API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.analysisservices.tabular), and Tabular Editor simply builds on top of this. In other words, whenever Microsoft makes any updates to the TOM, I only need to download the latest version of the API and recompile Tabular Editor. In other words, I do not have to manually manipulate the model metadata JSON - the API takes care of that.

Until something similar is available for the LSDL (ideally, some .NET library that can load the LSDL schema from JSON/XML/YAML into an object model and vice versa), please do not expect any improvements in this area.

**Comment by aaron-meyers** [comment]
> Hi @benito2313
> 
> There will not be better linguistic schema support in Tabular Editor, as long as Microsoft does not provide an API for programatically accessing the schema. Implementing my own API based on the linguistic schema specs would be a tremendous effort, and it would have to me maintained every time there is an update of the schema (which happens often!).
> 
> This is in contrast to the TOM itself, where [Microsoft provided an excellent API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.analysisservices.tabular), and Tabular Editor simply builds on top of this. In other words, whenever Microsoft makes any updates to the TOM, I only need to download the latest version of the API and recompile Tabular Editor. In other words, I do not have to manually manipulate the model metadata JSON - the API takes care of that.
> 
> Until something similar is available for the LSDL (ideally, some .NET library that can load the LSDL schema from JSON/XML/YAML into an object model and vice versa), please do not expect any improvements in this area.

This makes sense and I'll see what I can do about this. Of course, we have an API internally and besides just providing a nicer interface for working with the JSON it also implements version upgrade which I think would be helpful for Tabular Editor as well. Realistically I have no idea when we can get to exposing this but it's good to know there is interest :)

**Comment by otykier** [comment]
@aaron-meyers it’s been a while. Any updates on your last comment? :-)

**Comment by Rasenkantenstein** [comment]
@otykier that's indeed quite annoying then.
As a suggestion: According to this old post (https://powerbi.microsoft.com/en-us/blog/editing-q-a-linguistic-schemas/) it is possible to also import/export linguistic schemas from/into models. Given an existing yaml-file, would it be possible to import that into an existing model using tabular editor?

**Comment by Chris-Burgess-ACTGOV** [comment]
Hey all
anyone already started hitting up Microsoft for the synonym visibility?  - Is there a forum item i can add to?
Just dont want to split the wheel

Cheers
Chris

**Comment by aaron-meyers** [comment]
We're working out the details but I can at least say that we are working on a story for exposing Copilot tooling metadata (currently stored in the LinguisticMetadata in TOM/TMDL) in a form that works better with Git integration. I don't know if there's a forum item for this but we have the feedback internally and it's being planned.

As for LSDL JSON vs YAML - they are both based on the same LSDL JSON schema (at http://aka.ms/lsdlschema) and YAML is a syntactic superset of JSON so e.g. if you take the LSDL JSON from LinguisticMetadata and import it using the Import linguistic schema feature in PBI Desktop, it will work. The LSDL JSON schema supports both "concise" and "canonical" representations of some objects so if you convert the YAML output from PBI Desktop to JSON and compare it with the JSON in LinguisticMetadata you will see some differences, but both forms are "correct" and fully supported. We'll looking into simplifying this as part of the Copilot tooling effort.
